pkg/server/message_function_server.go

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018-Present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package server
+
+import (
+	"io"
+	"log"
+
+	"github.com/projectriff/shell-function-invoker/pkg/function"
+)
+
+type messageFunctionServer struct {
+	fnUri string
+}
+
+func New(fnUri string) *messageFunctionServer {
```

Comment:
It's idiomatic Go and is the common pattern in the standard library. It avoids tying the constructor function to the interface. Of course, `*messageFunctionServer` _is a_ `MessageFunctionServer`.

Reply:
And so it's tying it to the actual implementation? How is that better? 😕 
Or am I missing something?

Reply:
I tend to agree with you BTW. But playing devil's advocate, by returning a pointer, it will conform to whatever interface the user of `New` wants to define.

I will change it if you feel strongly, but I feel uncomfortable about breaking a common Go idiom as it will make the code look a bit strange to Go-literate newcomers. They might think it has been written by Java programmers...

Reply:
> I tend to agree with you BTW. But playing devil's advocate, by returning a pointer, it will conform to whatever interface the user of New wants to define.

I don't understand this. *We* own the contract of `New`, and it is that what is returned is a `MessageFunctionServer` (which doesn't preclude it to be something else as well). In that respect, golang is better than Java, because a Java construction for class `FooImpl` is named `FooImpl` and returns something that the caller sees is a `FooImpl` (although you only wanted to advertise it to be a `Foo` (interface))